published/external/xdpapi_helper.h

```diff
@@ -0,0 +1,279 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <xdpapi.h>
+
+#if defined(_KERNEL_MODE)
+
+#include <wdm.h>
+#include <xdp/object.h>
+
+static XDP_API_CLIENT _XdpApiContext = {0};
+static const XDP_API_PROVIDER_DISPATCH *_XdpApi;
+static const XDP_API_PROVIDER_BINDING_CONTEXT *_ProviderBindingContext;
+
+#define XDP_CREATE_HANDLE_WITH_RUNDOWN(FunctionCall, Object, ...)                               \
+    do {                                                                                        \
+        XDP_STATUS Status;                                                                      \
+        if (!ExAcquireRundownProtectionCacheAware(_XdpApiContext.RundownRef)) {                 \
+            return STATUS_DEVICE_NOT_READY;                                                     \
+        } else {                                                                                \
+            XDP_API_PROVIDER_BINDING_CONTEXT *ProviderBindingContext =                          \
+                (XDP_API_PROVIDER_BINDING_CONTEXT *)_ProviderBindingContext;                    \
+            Status = FunctionCall(ProviderBindingContext, __VA_ARGS__);                         \
+            if (NT_SUCCESS(Status)) {                                                           \
+                ((XDP_FILE_OBJECT_HEADER *)(Object))->RundownRef = _XdpApiContext.RundownRef;   \
+            } else {                                                                            \
+                ExReleaseRundownProtectionCacheAware(_XdpApiContext.RundownRef);                \
+            }                                                                                   \
+            return Status;                                                                      \
+        }                                                                                       \
+    } while (0)
+
+_IRQL_requires_(PASSIVE_LEVEL)
+NTSTATUS
+CxPlatXdpOpenApi(
+    _In_ UINT32 _XdpApiVersion,
+    _In_opt_ VOID *_ClientContext,
+    _In_opt_ XDP_API_ATTACH_FN *_ClientAttach,
+    _In_opt_ XDP_API_DETACH_FN *_ClientDetach,
+    _In_ const XDP_API_CLIENT_DISPATCH *_XdpApiClientDispatch,
+    _In_ const INT64 _TimeoutMs
+    )
+{
+    NTSTATUS Status;
+    const XDP_API_PROVIDER_DISPATCH *ProviderDispatch;
+    const XDP_API_PROVIDER_BINDING_CONTEXT *ProviderContext;
+
+    Status =
+        XdpOpenApi(
+            _XdpApiVersion,
+            _ClientContext,
+            _ClientAttach,
+            _ClientDetach,
+            _XdpApiClientDispatch,
+            _TimeoutMs,
+            &_XdpApiContext,
+            &ProviderDispatch,
+            &ProviderContext);
+    if (!NT_SUCCESS(Status)) {
+        return Status;
+    }
+
+    _XdpApi = (XDP_API_PROVIDER_DISPATCH *)ProviderDispatch;
+    _ProviderBindingContext = (XDP_API_PROVIDER_BINDING_CONTEXT *)ProviderContext;
+
+    return STATUS_SUCCESS;
+}
+
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID
+CxPlatXdpCloseApi(
```

Comment:
I don't think we should use the `CxPlat` prefix outside the microsoft/cxplat project - here we should use `Xdp` as the starting prefix, plus any more characters afterwards we think we need.

Reply:
I made it `XdpHlp` prefix to avoid conflict with normal APIs.